tests/app/authentication/test_authentication.py

```diff
@@ -1,478 +1,331 @@
 import time
 import uuid
-from datetime import datetime
-from unittest.mock import call
 
 import jwt
 import pytest
-from flask import current_app, json, request
-from freezegun import freeze_time
+from flask import current_app, g, request
 from notifications_python_client.authentication import create_jwt_token
 
-from app import api_user
+from app import db
 from app.authentication.auth import (
     GENERAL_TOKEN_ERROR_MESSAGE,
     AuthError,
-    requires_admin_auth,
+    _decode_jwt_token,
+    _get_auth_token,
+    _get_token_issuer,
     requires_auth,
+    requires_internal_auth,
 )
 from app.dao.api_key_dao import (
     expire_api_key,
     get_model_api_keys,
-    get_unsigned_secret,
     get_unsigned_secrets,
-    save_model_api_key,
 )
 from app.dao.services_dao import dao_fetch_service_by_id
-from app.models import KEY_TYPE_NORMAL, ApiKey
-from tests.conftest import set_config, set_config_values
+from tests.conftest import set_config_values
 
 
-@pytest.mark.parametrize('auth_fn', [requires_auth, requires_admin_auth])
-def test_should_not_allow_request_with_no_token(client, auth_fn):
-    request.headers = {}
-    with pytest.raises(AuthError) as exc:
-        auth_fn()
-    assert exc.value.short_message == 'Unauthorized: authentication token must be provided'
+@pytest.fixture
+def internal_jwt_token(notify_api):
+    with set_config_values(notify_api, {
+        'INTERNAL_CLIENT_API_KEYS': {
+            'myapp': ['myapp-secret'],
+        }
+    }):
+        yield create_jwt_token(
+            client_id='myapp',
+            secret='myapp-secret'
+        )
 
 
-@pytest.mark.parametrize('auth_fn', [requires_auth, requires_admin_auth])
-def test_should_not_allow_request_with_incorrect_header(client, auth_fn):
-    request.headers = {'Authorization': 'Basic 1234'}
-    with pytest.raises(AuthError) as exc:
-        auth_fn()
-    assert exc.value.short_message == 'Unauthorized: authentication bearer scheme must be used'
+def requires_myapp_auth():
+    requires_internal_auth('myapp')
 
 
-@pytest.mark.parametrize('auth_fn', [requires_auth, requires_admin_auth])
-def test_should_not_allow_request_with_incorrect_token(client, auth_fn):
-    request.headers = {'Authorization': 'Bearer 1234'}
-    with pytest.raises(AuthError) as exc:
-        auth_fn()
-    assert exc.value.short_message == GENERAL_TOKEN_ERROR_MESSAGE
+def create_custom_jwt_token(headers=None, payload=None, key=None):
```

Comment:
Optional: I'm not sure if you go down this rabbit hole, given the amount of refactoring already done in this PR, but I dislike the fact that in the python client we have

```
def create_jwt_token(secret, client_id):
    ....
    t = jwt.encode(payload=claims, key=secret, headers=headers)
    ....
```

and in this we have

```
def create_custom_jwt_token(headers=None, payload=None, key=None):
    # code copied from notifications_python_client.authentication.py::create_jwt_token
    headers = headers or {"typ": 'JWT', "alg": 'HS256'}
    return jwt.encode(payload=payload, key=key or str(uuid.uuid4()), headers=headers)
```

Explicitely, in one method we call key a key and in the other we call key a secret.

Ideally they would be called the same thing. JWT is often criticised for being confusing with lots of similar concepts, variable names, function names etc being hard to differentiate (and it can have a big impact if you get things confused with each other). However, I haven't thought long enough about whether this would be a quick win to fix or not (or even how you would fix it).

Feel free to do as you judge although I would err on the side of not including it in this PR if it was more than say 10 lines of code.

Reply:
Oh yes, that's annoying. I can see why we may have used "secret" elsewhere, and it's easier to go in that direction, so I've added added an extra commit to rename it except when we call the JWT code.